File: practice/src/main/java/com/decade/practice/web/rest/EventController.java
package com.decade.practice.web.rest;

import com.decade.practice.data.repositories.UserRepository;
import com.decade.practice.media.ImageStore;
import com.decade.practice.media.MediaStore;
import com.decade.practice.models.domain.embeddable.ChatIdentifier;
import com.decade.practice.models.domain.embeddable.ImageSpec;
import com.decade.practice.models.domain.entity.*;
import com.decade.practice.presence.UserPresenceService;
import com.decade.practice.usecases.ChatOperations;
import com.decade.practice.usecases.EventOperations;
import com.decade.practice.utils.CacheUtils;
import com.decade.practice.utils.ChatUtils;
import com.decade.practice.utils.EventUtils;
import com.decade.practice.utils.ImageUtils;
import jakarta.persistence.EntityNotFoundException;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.awt.image.BufferedImage;
import java.io.IOException;
import java.net.URI;
import java.security.Principal;
import java.util.List;
import java.util.UUID;

import static java.lang.Math.min;

@RestController
@RequestMapping
public class EventController {

        private final EventOperations eventOperations;
        private final ChatOperations chatOperations;
        private final UserPresenceService presenceService;
        private final UserRepository userRepo;
        private final ImageStore imageStore;
        private final MediaStore mediaStore;

        public EventController(
                EventOperations eventOperations,
                ChatOperations chatOperations,
                UserPresenceService presenceService,
                UserRepository userRepo,
                ImageStore imageStore,
                MediaStore mediaStore) {
                this.eventOperations = eventOperations;
                this.chatOperations = chatOperations;
                this.presenceService = presenceService;
                this.userRepo = userRepo;
                this.imageStore = imageStore;
                this.mediaStore = mediaStore;
        }


        @PostMapping(path = "/events", consumes = {MediaType.APPLICATION_JSON_VALUE})
        @ResponseStatus(HttpStatus.CREATED)
        public ChatEvent createEvent(
                Principal principal,
                @RequestBody @Valid ChatEvent event) {
                User user = userRepo.getByUsername(principal.getName());
                presenceService.set(user);
                return eventOperations.createAndSend(user, event);
        }


        @PostMapping(path = "/events", consumes = {MediaType.MULTIPART_FORM_DATA_VALUE}, headers = "X-File-Type=image")
        @ResponseStatus(HttpStatus.CREATED)
        public ImageEvent createEvent(
                Principal principal,
                @RequestPart("event") @Valid ImageEvent event,
                @RequestPart("file") MultipartFile file
        ) throws EntityNotFoundException, IOException {
                int maxWidth = min(event.getImage().getWidth(), ImageSpec.DEFAULT_WIDTH);
                int maxHeight = min(event.getImage().getHeight(), ImageSpec.DEFAULT_HEIGHT);
                BufferedImage image = ImageUtils.crop(file.getInputStream(), maxWidth, maxHeight);

                event.setImage(imageStore.save(image));
                try {
                        User user = userRepo.getByUsername(principal.getName());
                        presenceService.set(user);
                        return eventOperations.createAndSend(user, event);
                } catch (Exception e) {
                        imageStore.remove(URI.create(event.getImage().getUri()));
                        throw e;
                }
        }

        @PostMapping(path = "/events", consumes = {MediaType.MULTIPART_FORM_DATA_VALUE}, headers = "X-File-Type=file")
        @ResponseStatus(HttpStatus.CREATED)
        public FileEvent createEvent(
                Principal principal,
                @RequestPart("event") @Valid FileEvent event,
                @RequestPart("file") MultipartFile file
        ) throws EntityNotFoundException, IOException {
                URI uri = mediaStore.save(file.getResource(), UUID.randomUUID().toString());
                event.setMediaUrl(uri.toString());
                try {
                        User user = userRepo.getByUsername(principal.getName());
                        presenceService.set(user);
                        return eventOperations.createAndSend(user, event);
                } catch (Exception e) {
                        mediaStore.remove(uri);
                        throw e;
                }
        }

        @GetMapping("/chats/{chatIdentifier}/events")
        public ResponseEntity<List<ChatEvent>> listEvents(
                Principal principal,
                @PathVariable @Validated ChatIdentifier chatIdentifier,
                @RequestParam int atVersion
        ) throws EntityNotFoundException {
                Chat chat = chatOperations.getOrCreateChat(chatIdentifier);
                User user = ChatUtils.inspectOwner(chat, principal.getName());
                List<ChatEvent> events = eventOperations.findByOwnerAndChatAndEventVersionLessThanEqual(
                        user, chat, atVersion, EventUtils.EVENT_VERSION_LESS_THAN_EQUAL);

                return ResponseEntity.ok()
                        .header("Vary", "Cookie", "Authorization")
                        .cacheControl(CacheUtils.ONE_MONTHS)
                        .body(events);
        }

        @GetMapping("/users/me/events")
        public ResponseEntity<List<ChatEvent>> listEvents(
                Principal principal,
                @RequestParam int atVersion
        ) throws EntityNotFoundException {
                User owner = userRepo.getByUsername(principal.getName());
                List<ChatEvent> eventList = eventOperations.findByOwnerAndEventVersionLessThanEqual(
                        owner, atVersion, EventUtils.EVENT_VERSION_LESS_THAN_EQUAL);

                return ResponseEntity.ok()
                        .header("Vary", "Cookie", "Authorization")
                        .cacheControl(CacheUtils.ONE_MONTHS)
                        .body(eventList);
        }
}

File: practice/src/main/java/com/decade/practice/endpoints/chat/EventController.kt
package com.decade.practice.endpoints.chat

import com.decade.practice.core.ChatOperations
import com.decade.practice.core.OnlineStatistic
import com.decade.practice.database.repository.EventRepository
import com.decade.practice.database.repository.UserRepository
import com.decade.practice.database.transaction.ChatEventStore
import com.decade.practice.image.ImageStore
import com.decade.practice.model.domain.embeddable.ChatIdentifier
import com.decade.practice.model.domain.embeddable.ImageSpec.Companion.DEFAULT_HEIGHT
import com.decade.practice.model.domain.embeddable.ImageSpec.Companion.DEFAULT_WIDTH
import com.decade.practice.model.domain.entity.*
import com.decade.practice.util.EventPageUtils
import com.decade.practice.util.ImageUtils
import com.decade.practice.util.inspectOwner
import com.decade.practice.websocket.MQ_DESTINATION
import jakarta.persistence.EntityNotFoundException
import jakarta.validation.Valid
import org.springframework.dao.DataIntegrityViolationException
import org.springframework.http.CacheControl
import org.springframework.http.ResponseEntity
import org.springframework.messaging.simp.SimpMessagingTemplate
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.*
import org.springframework.web.multipart.MultipartFile
import java.io.IOException
import java.net.URI
import java.util.concurrent.TimeUnit
import kotlin.math.min


@RestController
@RequestMapping("/message")
class EventController(
      private val chatEventStore: ChatEventStore,
      private val template: SimpMessagingTemplate,
      private val chatOperations: ChatOperations,
      private val evenRepo: EventRepository,
      private val onlineStat: OnlineStatistic,
      private val userRepo: UserRepository,
      private val imageStore: ImageStore,
) {

      @Throws(EntityNotFoundException::class)
      private fun <E : ChatEvent> pushEvent(
            sender: User,
            @RequestBody event: E
      ): E {
            onlineStat.set(sender)
            event.sender = sender
            try {
                  val saved = chatEventStore.save(event)
                  saved.forEach {
                        template.convertAndSendToUser(
                              it.owner.username,
                              MQ_DESTINATION,
                              it
                        )
                  }
                  return saved.find {
                        it.localId == event.localId
                  } as E
            } catch (e: DataIntegrityViolationException) {
                  // record already sent
                  return evenRepo.getByLocalId(event.localId) as E
            }
      }

      @PostMapping("/seen")
      @Throws(EntityNotFoundException::class)
      fun pushEvent(
            @AuthenticationPrincipal(expression = "name") username: String,
            @RequestBody @Valid record: SeenEvent
      ): SeenEvent = pushEvent(userRepo.getByUsername(username), record)


      @PostMapping("/text")
      @Throws(EntityNotFoundException::class)
      fun pushEvent(
            @AuthenticationPrincipal(expression = "name") username: String,
            @RequestBody @Valid event: TextEvent
      ): TextEvent = pushEvent(userRepo.getByUsername(username), event)


      @PostMapping("/icon")
      @Throws(EntityNotFoundException::class)
      fun pushEvent(
            @AuthenticationPrincipal(expression = "name") username: String,
            @RequestBody @Valid event: IconEvent
      ): IconEvent = pushEvent(userRepo.getByUsername(username), event)


      @PostMapping("/image")
      @Throws(EntityNotFoundException::class, IOException::class)
      fun pushEvent(
            @AuthenticationPrincipal(expression = "name") username: String,
            @RequestPart("event") @Valid event: ImageEvent,
            @RequestPart("file") file: MultipartFile
      ): ImageEvent {
            val maxWidth = min(event.image.width, DEFAULT_WIDTH)
            val maxHeight = min(event.image.height, DEFAULT_HEIGHT)
            val image = ImageUtils.crop(file.inputStream, maxWidth, maxHeight)

            event.image = imageStore.save(image)
            try {
                  return pushEvent(userRepo.getByUsername(username), event)
            } catch (e: Exception) {
                  imageStore.remove(URI(event.image.uri))
                  throw e
            }
      }

      @GetMapping("/chat")
      @Throws(EntityNotFoundException::class)
      fun pullEvents(
            @AuthenticationPrincipal(expression = "name") username: String,
            @RequestParam @Validated identifier: ChatIdentifier,
            @RequestParam atVersion: Int
      ): ResponseEntity<List<ChatEvent>> {

            val chat = chatOperations.getOrCreateChat(identifier)
            val user = chat.inspectOwner(username)
            val page = EventPageUtils.pageEvent
            val cacheControl = CacheControl.maxAge(30, TimeUnit.DAYS)
                  .cachePublic()
            val events = evenRepo.findByOwnerAndChatAndEventVersionLessThanEqual(user, chat, atVersion, page)
            return ResponseEntity.ok()
                  .varyBy("Cookie", "Authorization")
                  .cacheControl(cacheControl)
                  .body(events)
      }

      @GetMapping
      @Throws(EntityNotFoundException::class)
      fun pullEvents(
            @AuthenticationPrincipal(expression = "name") username: String,
            @RequestParam atVersion: Int
      ): ResponseEntity<List<ChatEvent>> {

            val owner = userRepo.getByUsername(username)
            val page = EventPageUtils.pageEvent
            val cacheControl = CacheControl.maxAge(30, TimeUnit.DAYS)
                  .cachePublic()
            val eventList = evenRepo.findByOwnerAndEventVersionLessThanEqual(owner, atVersion, page)
            return ResponseEntity.ok()
                  .varyBy("Cookie", "Authorization")
                  .cacheControl(cacheControl)
                  .body(eventList)
      }
}
